### PR TITLE
Use latest lindi, pynwb, and hdmf with aws_region

### PR DESCRIPTION
**Pull request by rly**
I updated the environment file. Please remake your conda environments after merging. 

I added a couple of read remote LINDI file test cases. I think the local lindi test cases still apply, though we should update those to point to the dandiset with the test NWB files that we want to use.  

I'm pointing the environment to branches of HDMF and PyNWB that support the `aws_region` argument to `NWBHDF5IO` so that we can specify the AWS region in ROS3 tests.

Also we no longer need to say `load_namespaces=True`. That is now the default.

### CODE REVIEW

**Comment by rly** [comment]
```bash
nwb_benchmarks run --bench NWBLindiFileReadRemoteReferenceFileSystemBenchmark
nwb_benchmarks run --bench NWBLindiFileCreateLocalReferenceFileSystemBenchmark
nwb_benchmarks run --bench LindiFileReadLocalReferenceFileSystemBenchmark
```
These ran successfully. I have not yet run a full suite of benchmarks yet though.

**Comment by rly** [comment]
I defined remote benchmarks as such:
```python
lindi_remote_rfs_parameter_cases = dict(
    EcephysTestCase=dict(
        s3_url=get_s3_url(
            is_staging=True,
            dandiset_id="213889",
            dandi_path="sub-IBL-ecephys/sub-IBL-ecephys_ses-3e7ae7c0_desc-18000000-frames-13653-by-384-chunking.lindi.json",
        ),
        object_name="ElectricalSeriesAp",
        slice_range=(slice(0, 30_000), slice(0, 384)),
    ),
    OphysTestCase=dict(
        s3_url=get_s3_url(
            is_staging=True,
            dandiset_id="213889",
            dandi_path="sub-R6_ses-20200206T210000_behavior+ophys/sub-R6_ses-20200206T210000_behavior+ophys.lindi.json",
        ),
        object_name="TwoPhotonSeries",
        slice_range=(slice(0, 3), slice(0, 796), slice(0, 512)),
    ),
    IcephysTestCase=dict(
        s3_url=get_s3_url(
            is_staging=True,
            dandiset_id="213889",
            dandi_path="sub-1214579789_ses-1214621812_icephys/sub-1214579789_ses-1214621812_icephys.lindi.json",
        ),
        object_name="data_00002_AD0",
        slice_range=(slice(0, 30_000),),
    ),
```

I thought that this would create three test cases. Instead, it creates 9 - the cross of all the parameters. But all of the unintended combinations will fail. I think this is an issue within `BaseBenchmark`. @CodyCBakerPhD is this behavior correct? and if so, how do I create three test cases instead of the cross of all these options?

File: src/nwb_benchmarks/core/_dandi.py [hunk 1]
```diff
@@ -16,7 +16,8 @@ def get_s3_url(dandiset_id: str, dandi_path: str) -> str:
     """
     assert len(dandiset_id) == 6, f"The specified 'dandiset_id' ({dandiset_id}) should be the six-digit identifier."
 
-    client = DandiAPIClient()
+    api_url = "https://api-staging.dandiarchive.org/api" if is_staging else "https://api.dandiarchive.org/api"
```
Comment (CodyCBakerPhD) [inline src/nwb_benchmarks/core/_dandi.py:19]:
```suggestion
    api_url = "https://api-staging.dandiarchive.org/api" if int(dandiset_id) >= 200_000 else "https://api.dandiarchive.org/api"
```

No need for flag, you can tell by the ID

File: src/nwb_benchmarks/core/_dandi.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 from dandi.dandiapi import DandiAPIClient
 
 
-def get_s3_url(dandiset_id: str, dandi_path: str) -> str:
+def get_s3_url(dandiset_id: str, dandi_path: str, is_staging: bool = False) -> str:
```
Comment (CodyCBakerPhD) [inline src/nwb_benchmarks/core/_dandi.py:4]:
```suggestion
def get_s3_url(dandiset_id: str, dandi_path: str) -> str:
```

**Comment by CodyCBakerPhD** [comment]
@rly I get

```python
Traceback (most recent call last):
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\asv\benchmark.py", line 68, in <module>
                 main()
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\asv\benchmark.py", line 60, in main
                 commands[mode](args)
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\asv_runner\run.py", line 72, in _run
                 result = benchmark.do_run()
                          ^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\asv_runner\benchmarks\_base.py", line 661, in do_run
                 return self.run(*self._current_params)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\asv_runner\benchmarks\time.py", line 165, in run
                 samples, number = self.benchmark_timing(
                                   ^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\asv_runner\benchmarks\time.py", line 258, in benchmark_timing
                 timing = timer.timeit(number)
                          ^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\timeit.py", line 180, in timeit
                 timing = self.inner(it, self.timer)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^
               File "<timeit-src>", line 6, in inner
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\asv_runner\benchmarks\time.py", line 90, in func
                 self.func(*param)
               File "D:\GitHub\nwb_benchmarks\src\nwb_benchmarks\benchmarks\time_remote_file_reading.py", line 235, in time_read_lindi_nwbfile
                 self.nwbfile, self.io, self.client = read_hdf5_nwbfile_lindi(rfs=s3_url)
                                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
               File "D:\GitHub\nwb_benchmarks\src\nwb_benchmarks\core\_streaming.py", line 227, in read_hdf5_nwbfile_lindi
                 nwbfile = io.read()
                           ^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\utils.py", line 668, in func_call
                 return func(args[0], **pargs)
                        ^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\pynwb\__init__.py", line 304, in read
                 file = super().read(**kwargs)
                        ^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\backends\hdf5\h5tools.py", line 482, in read
                 return super().read(**kwargs)
                        ^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\utils.py", line 668, in func_call
                 return func(args[0], **pargs)
                        ^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\backends\io.py", line 60, in read
                 container = self.__manager.construct(f_builder)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\utils.py", line 668, in func_call
                 return func(args[0], **pargs)
                        ^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\build\manager.py", line 284, in construct
                 result = self.__type_map.construct(builder, self, None)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\utils.py", line 668, in func_call
                 return func(args[0], **pargs)
                        ^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\build\manager.py", line 795, in construct
                 return obj_mapper.construct(builder, build_manager, parent)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\utils.py", line 668, in func_call
                 return func(args[0], **pargs)
                        ^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\build\objectmapper.py", line 1228, in construct
                 subspecs = self.__get_subspec_values(builder, self.spec, manager)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\build\objectmapper.py", line 1157, in __get_subspec_values
                 self.__get_sub_builders(groups, spec.groups, manager, ret)
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\build\objectmapper.py", line 1208, in __get_sub_builders
                 ret.update(self.__get_subspec_values(sub_builder, subspec, manager))
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\build\objectmapper.py", line 1157, in __get_subspec_values
                 self.__get_sub_builders(groups, spec.groups, manager, ret)
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\build\objectmapper.py", line 1186, in __get_sub_builders
                 for parent_dt in manager.namespace_catalog.get_hierarchy(ns, dt):
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\utils.py", line 668, in func_call
                 return func(args[0], **pargs)
                        ^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\theac\anaconda3\envs\nwb_benchmarks_created_5_18_2024\Lib\site-packages\hdmf\spec\namespace.py", line 328, in get_hierarchy
                 raise KeyError("'%s' not a namespace" % namespace)
             KeyError: "'ndx-mies' not a namespace"
```

when running `time_remote_file_reading.NWBLindiFileReadRemoteReferenceFileSystemBenchmark.time_read_lindi_nwbfile`

FYI

File: src/nwb_benchmarks/core/_dandi.py (hunk 1 above)
Comment (CodyCBakerPhD) [reply src/nwb_benchmarks/core/_dandi.py:19]: Moved this to #61

**Comment by CodyCBakerPhD** [comment]
> @CodyCBakerPhD is this behavior correct? and if so, how do I create three test cases instead of the cross of all these options?

Well, that's cool...

First time running that parameterized pattern and it's revealed that that is indeed the intended behavior of ASV: https://asv.readthedocs.io/en/v0.6.1/writing_benchmarks.html#parameterized-benchmarks

There's not an easy way I can think of at the moment to resolve this at the `parameter_case` or `Basebenchmark` level (attempts to use `parameterized` package cause issues in benchmark discovery phase)

Only way will be to break up the multilevel dict-of-dicts to separate benchmarks. I'll do that in my PR

File: src/nwb_benchmarks/core/_base_benchmark.py [hunk 3]
```diff
@@ -27,19 +31,43 @@ class BaseBenchmark:
...
+                if len(cls.param_names) > 1
                 else [
                     parameter
-                    for parameter_case in self.parameter_cases.values()
+                    for parameter_case in cls.parameter_cases.values()
                     for parameter in parameter_case.values()
                 ]
             )
+
+            # ASV automatically forms a cartesian product over all params
```
Comment (CodyCBakerPhD) [inline src/nwb_benchmarks/core/_base_benchmark.py:53]: Or, I guess I had already pushed the fix to the params thing on the other branch. Now #63 has a minor residual fix to parsing the results from the corrected intermediate results files as well as some more results

**Comment by rly** [comment]
This fix in lindi makes the lindi ophys test case work. https://github.com/NeurodataWithoutBorders/lindi/pull/71

File: environments/nwb_benchmarks.yaml [hunk 4]
```diff
@@ -9,15 +9,14 @@ dependencies:
...
       - dandi
       - fsspec
       - requests
       - aiohttp
       - remfile
       - pyshark
-      - lindi
-      - hdmf @ git+https://github.com/hdmf-dev/hdmf.git@9b3282aa4999d2922f003f1b79ec7458ea3ddc5e  # required until PyNWB propagates changes
-      #- hdmf @ git+https://github.com/hdmf-dev/hdmf.git@expose_aws_region  # required until region fix is released
+      - lindi>=0.3.5
```
Comment (CodyCBakerPhD) [inline environments/nwb_benchmarks.yaml:18]:
```suggestion
      - lindi>=0.3.6
```

Required for latest fix

**Review by CodyCBakerPhD — APPROVED** [review]

**Pull request merged**